File: trunk/java/UMLGraph/src/org/daisy/urakawa/undo/CommandManager.java
package org.daisy.urakawa.undo;

import org.daisy.urakawa.exception.MethodParameterIsNullException;

/**
 * <p>
 * The Class realizing this interface must store the history of Commands and
 * maintain the current Undo/Redo state (e.g. pointer in the dynamic stack of
 * Commands).
 * </p>
 * 
 * @depend - "Composition\n(history)" 0..n org.daisy.urakawa.undo.Command
 * @leafInterface see {@link org.daisy.urakawa.LeafInterface}
 * @see org.daisy.urakawa.LeafInterface
 * @stereotype OptionalLeafInterface
 */
public interface CommandManager {
	/**
	 * <p>
	 * Returns a human-readable name for the next Undoable Command.
	 * </p>
	 * 
	 * @return cannot be null or empty string.
	 * @tagvalue Exceptions "CannotUndo"
	 * @throws CannotUndoException
	 * @see #canUndo()
	 * @see org.daisy.urakawa.undo.Command#getUnExecuteShortDescription()
	 */
	public String getUndoShortDescription() throws CannotUndoException;

	/**
	 * <p>
	 * undoes the last executed Command
	 * </p>
	 * 
	 * @tagvalue Exceptions "CannotUndo"
	 * @throws CannotUndoException
	 * @see #canUndo()
	 * @see org.daisy.urakawa.undo.Command#unExecute() ()
	 */
	public void undo() throws CannotUndoException;

	/**
	 * <p>
	 * Returns a human-readable name for the next Redoable Command
	 * </p>
	 * 
	 * @return cannot be null or empty string.
	 * @tagvalue Exceptions "CannotRedo"
	 * @throws CannotRedoException
	 * @see #canRedo()
	 * @see org.daisy.urakawa.undo.Command#getExecuteShortDescription()
	 */
	public String getRedoShortDescription() throws CannotRedoException;

	/**
	 * <p>
	 * redoes the last undone Command
	 * </p>
	 * 
	 * @tagvalue Exceptions "CannotRedo"
	 * @throws CannotRedoException
	 * @see #canRedo()
	 * @see org.daisy.urakawa.undo.Command#execute()
	 */
	public void redo() throws CannotRedoException;

	/**
	 * <p>
	 * Executes and registers the given Command in the history, deleting all
	 * following undone Commands in the history (on the "right hand side"), if
	 * any.
	 * </p>
	 * <p>
	 * In some special cases (e.g. user typing text letter by letter, but
	 * undo/redo applies to full word or sentence), this method my take the
	 * responsibility to automatically merge a series of Commands into a
	 * CompositeCommand.
	 * </p>
	 * 
	 * @param command
	 *            the Command to register in the history and to execute.
	 * @tagvalue Exceptions "MethodParameterIsNull"
	 * @throws MethodParameterIsNullException
	 *             NULL method parameters are forbidden
	 */
	public void execute(Command command) throws MethodParameterIsNullException;

	/**
	 * <p>
	 * Tests whether it's possible to undo.
	 * </p>
	 * 
	 * @return false if the history is empty, otherwise true if the last
	 *         executed Command (via "execute()" or "redo()")) is undoable.
	 * @see org.daisy.urakawa.undo.Command#canUnExecute()
	 */
	public boolean canUndo();

	/**
	 * <p>
	 * Tests whether it's possible to redo.
	 * </p>
	 * 
	 * @return false if the history is empty, otherwise true if the last undone
	 *         Command (via "undo()")) is redoable.
	 */
	public boolean canRedo();
}

File: csharp/UrakawaToolkit/undo/CommandManager.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace urakawa.undo
{
    /// <summary>
    /// The command manager.
    /// </summary>
    public class CommandManager
    {
        private Stack<ICommand> mUndoStack;  // stack of commands to undo
        private Stack<ICommand> mRedoStack;  // stack of commands to redo

        /// <summary>
        /// Create an empty command manager.
        /// </summary>
        public CommandManager()
        {
            mUndoStack = new Stack<ICommand>();
            mRedoStack = new Stack<ICommand>();
        }

        /// <summary>
        /// Get the name of the next undo command.
        /// </summary>
        /// <exception cref="CannotUndoException">Thrown when there is no command to undo.</exception>
        public string getUndoShortDescription()
        {
            if (mUndoStack.Count == 0) throw new CannotUndoException("There is no command to undo.");
            return mUndoStack.Peek().getUnExecuteShortDescription();
        }

        /// <summary>
        /// Undo the last executed command.
        /// </summary>
        /// <exception cref="CannotUndoException">Thrown when there is no command to undo.</exception>
        public void undo()
        {
            if (mUndoStack.Count == 0) throw new CannotUndoException("There is no command to undo.");
            mUndoStack.Peek().unExecute();
            mRedoStack.Push(mUndoStack.Pop());
        }

        /// <summary>
        /// Get the name of the next redo command.
        /// </summary>
        /// <exception cref="CannotRedoException">Thrown when there is no command to redo.</exception>
        public string getRedoShortDescription()
        {
            if (mRedoStack.Count == 0) throw new CannotRedoException("There is no command to redo.");
            return mRedoStack.Peek().getExecuteShortDescription();
        }

        /// <summary>
        /// Redo the last unexecuted command.
        /// </summary>
        /// <exception cref="CannotRedoException">Thrown when there is no command to redo.</exception>
        public void redo()
        {
            if (mRedoStack.Count == 0) throw new CannotRedoException("There is no command to redo.");
            ICommand command = mRedoStack.Peek();
            mUndoStack.Push(mRedoStack.Pop());
            command.execute();
        }

        /// <summary>
        /// Execute and register the given command in the undo history and clear the redo history.
        /// </summary>
        /// <exception cref="exception.MethodParameterIsNullException">Thrown when a null command is given.</exception>
        public void execute(ICommand command)
        {
            if (command == null) throw new exception.MethodParameterIsNullException("Command cannot be null.");
            mUndoStack.Push(command);
            mRedoStack.Clear();
            command.execute();
        }

        /// <summary>
        /// Return true if the undo is history is non-empty.
        /// </summary>
        public bool canUndo()
        {
            return mUndoStack.Count > 0;
        }

        /// <summary>
        /// Return true if the redo is history is non-empty.
        /// </summary>
        public bool canRedo()
        {
            return mRedoStack.Count > 0;
        }
    }
}
